projects/libarchive-3.7.4/doc/man/cpio.5
.TH CPIO 5 "December 23, 2011" ""
.SH NAME
.ad l
\fB\%cpio\fP
\- format of cpio archive files
.SH DESCRIPTION
.ad l
The
\fB\%cpio\fP
archive format collects any number of files, directories, and other
file system objects (symbolic links, device nodes, etc.) into a single
stream of bytes.
.SS General Format
Each file system object in a
\fB\%cpio\fP
archive comprises a header record with basic numeric metadata
followed by the full pathname of the entry and the file data.
The header record stores a series of integer values that generally
follow the fields in
\fIstruct\fP stat.
(See
\fBstat\fP(2)
for details.)
The variants differ primarily in how they store those integers
(binary, octal, or hexadecimal).
The header is followed by the pathname of the
entry (the length of the pathname is stored in the header)
and any file data.
The end of the archive is indicated by a special record with
the pathname
``TRAILER!!!''.
.SS PWB format
The PWB binary
\fB\%cpio\fP
format is the original format, when cpio was introduced as part of the
Programmer's Work Bench system, a variant of 6th Edition UNIX.  It
stores numbers as 2-byte and 4-byte binary values.
Each entry begins with a header in the following format:
.PP
.RS 4
.nf
struct header_pwb_cpio {
        short   h_magic;
        short   h_dev;
        short   h_ino;
        short   h_mode;
        short   h_uid;
        short   h_gid;
        short   h_nlink;
        short   h_majmin;
        long    h_mtime;
        short   h_namesize;
        long    h_filesize;
};
.RE
.PP
The
\fIshort\fP
fields here are 16-bit integer values, while the
\fIlong\fP
fields are 32 bit integers.  Since PWB UNIX, like the 6th Edition UNIX
it was based on, only ran on PDP-11 computers, they
are in PDP-endian format, which has little-endian shorts, and
big-endian longs.  That is, the long integer whose hexadecimal
representation is 0x12345678 would be stored in four successive bytes
as 0x34, 0x12, 0x78, 0x56.
The fields are as follows:
.RS 5
.TP
\fIh_magic\fP
The integer value octal 070707.
.TP
\fIh_dev\fP, \fIh_ino\fP
The device and inode numbers from the disk.
These are used by programs that read
\fB\%cpio\fP
archives to determine when two entries refer to the same file.
Programs that synthesize
\fB\%cpio\fP
archives should be careful to set these to distinct values for each entry.
.TP
\fIh_mode\fP
The mode specifies both the regular permissions and the file type, and
it also holds a couple of bits that are irrelevant to the cpio format,
because the field is actually a raw copy of the mode field in the inode
representing the file.  These are the IALLOC flag, which shows that
the inode entry is in use, and the ILARG flag, which shows that the
file it represents is large enough to have indirect blocks pointers in
the inode.
The mode is decoded as follows:
.PP
.RS 5
.TP
0100000
IALLOC flag - irrelevant to cpio.
.TP
0060000
This masks the file type bits.
.TP
0040000
File type value for directories.
.TP
0020000
File type value for character special devices.
.TP
0060000
File type value for block special devices.
.TP
0010000
ILARG flag - irrelevant to cpio.
.TP
0004000
SUID bit.
.TP
0002000
SGID bit.
.TP
0001000
Sticky bit.
.TP
0000777
The lower 9 bits specify read/write/execute permissions
for world, group, and user following standard POSIX conventions.
.RE
.TP
\fIh_uid\fP, \fIh_gid\fP
The numeric user id and group id of the owner.
.TP
\fIh_nlink\fP
The number of links to this file.
Directories always have a value of at least two here.
Note that hardlinked files include file data with every copy in the archive.
.TP
\fIh_majmin\fP
For block special and character special entries,
this field contains the associated device number, with the major
number in the high byte, and the minor number in the low byte.
For all other entry types, it should be set to zero by writers
and ignored by readers.
.TP
\fIh_mtime\fP
Modification time of the file, indicated as the number
of seconds since the start of the epoch,
00:00:00 UTC January 1, 1970.
.TP
\fIh_namesize\fP
The number of bytes in the pathname that follows the header.
This count includes the trailing NUL byte.
.TP
\fIh_filesize\fP
The size of the file.  Note that this archive format is limited to 16
megabyte file sizes, because PWB UNIX, like 6th Edition, only used
an unsigned 24 bit integer for the file size internally.
.RE
.PP
The pathname immediately follows the fixed header.
If
\fBh_namesize\fP
is odd, an additional NUL byte is added after the pathname.
The file data is then appended, again with an additional NUL
appended if needed to get the next header at an even offset.
.PP
Hardlinked files are not given special treatment;
the full file contents are included with each copy of the
file.
.SS New Binary Format
The new binary
\fB\%cpio\fP
format showed up when cpio was adopted into late 7th Edition UNIX.
It is exactly like the PWB binary format, described above, except for
three changes:
.PP
First, UNIX now ran on more than one hardware type, so the endianness
of 16 bit integers must be determined by observing the magic number at
the start of the header.  The 32 bit integers are still always stored
with the most significant word first, though, so each of those two, in
the struct shown above, was stored as an array of two 16 bit integers,
in the traditional order.  Those 16 bit integers, like all the others
in the struct, were accessed using a macro that byte swapped them if
necessary.
.PP
Next, 7th Edition had more file types to store, and the IALLOC and ILARG
flag bits were re-purposed to accommodate these.  The revised use of the
various bits is as follows:
.PP
.RS 5
.TP
0170000
This masks the file type bits.
.TP
0140000
File type value for sockets.
.TP
0120000
File type value for symbolic links.
For symbolic links, the link body is stored as file data.
.TP
0100000
File type value for regular files.
.TP
0060000
File type value for block special devices.
.TP
0040000
File type value for directories.
.TP
0020000
File type value for character special devices.
.TP
0010000
File type value for named pipes or FIFOs.
.TP
0004000
SUID bit.
.TP
0002000
SGID bit.
.TP
0001000
Sticky bit.
.TP
0000777
The lower 9 bits specify read/write/execute permissions
for world, group, and user following standard POSIX conventions.
.RE
.PP
Finally, the file size field now represents a signed 32 bit integer in
the underlying file system, so the maximum file size has increased to
2 gigabytes.
.PP
Note that there is no obvious way to tell which of the two binary
formats an archive uses, other than to see which one makes more
sense.  The typical error scenario is that a PWB format archive
unpacked as if it were in the new format will create named sockets
instead of directories, and then fail to unpack files that should
go in those directories.  Running
\fIbsdcpio\fP -itv
on an unknown archive will make it obvious which it is: if it's
PWB format, directories will be listed with an 's' instead of
a 'd' as the first character of the mode string, and the larger
files will have a '?' in that position.
.SS Portable ASCII Format
Version 2 of the Single UNIX Specification (``SUSv2'')
standardized an ASCII variant that is portable across all
platforms.
It is commonly known as the
``old character''
format or as the
``odc''
format.
It stores the same numeric fields as the old binary format, but
represents them as 6-character or 11-character octal values.
.PP
.RS 4
.nf
struct cpio_odc_header {
        char    c_magic[6];
        char    c_dev[6];
        char    c_ino[6];
        char    c_mode[6];
        char    c_uid[6];
        char    c_gid[6];
        char    c_nlink[6];
        char    c_rdev[6];
        char    c_mtime[11];
        char    c_namesize[6];
        char    c_filesize[11];
};
.RE
.PP
The fields are identical to those in the new binary format.
The name and file body follow the fixed header.
Unlike the binary formats, there is no additional padding
after the pathname or file contents.
If the files being archived are themselves entirely ASCII, then
the resulting archive will be entirely ASCII, except for the
NUL byte that terminates the name field.
.SS New ASCII Format
The "new" ASCII format uses 8-byte hexadecimal fields for
all numbers and separates device numbers into separate fields
for major and minor numbers.
.PP
.RS 4
.nf
struct cpio_newc_header {
        char    c_magic[6];
        char    c_ino[8];
        char    c_mode[8];
        char    c_uid[8];
        char    c_gid[8];
        char    c_nlink[8];
        char    c_mtime[8];
        char    c_filesize[8];
        char    c_devmajor[8];
        char    c_devminor[8];
        char    c_rdevmajor[8];
        char    c_rdevminor[8];
        char    c_namesize[8];
        char    c_check[8];
};
.RE
.PP
Except as specified below, the fields here match those specified
for the new binary format above.
.RS 5
.TP
\fImagic\fP
The string
``070701''.
.TP
\fIcheck\fP
This field is always set to zero by writers and ignored by readers.
See the next section for more details.
.RE
.PP
The pathname is followed by NUL bytes so that the total size
of the fixed header plus pathname is a multiple of four.
Likewise, the file data is padded to a multiple of four bytes.
Note that this format supports only 4 gigabyte files (unlike the
older ASCII format, which supports 8 gigabyte files).
.PP
In this format, hardlinked files are handled by setting the
filesize to zero for each entry except the first one that
appears in the archive.
.SS New CRC Format
The CRC format is identical to the new ASCII format described
in the previous section except that the magic field is set
to
``070702''
and the
\fIcheck\fP
field is set to the sum of all bytes in the file data.
This sum is computed treating all bytes as unsigned values
and using unsigned arithmetic.
Only the least-significant 32 bits of the sum are stored.
.SS HP variants
The
\fB\%cpio\fP
implementation distributed with HPUX used XXXX but stored
device numbers differently XXX.
.SS Other Extensions and Variants
Sun Solaris uses additional file types to store extended file
data, including ACLs and extended attributes, as special
entries in cpio archives.
.PP
XXX Others? XXX
.SH SEE ALSO
.ad l
\fBcpio\fP(1),
\fBtar\fP(5)
.SH STANDARDS
.ad l
The
\fB\%cpio\fP
utility is no longer a part of POSIX or the Single Unix Standard.
It last appeared in
Version 2 of the Single UNIX Specification (``SUSv2'').
It has been supplanted in subsequent standards by
\fBpax\fP(1).
The portable ASCII format is currently part of the specification for the
\fBpax\fP(1)
utility.
.SH HISTORY
.ad l
The original cpio utility was written by Dick Haight
while working in AT&T's Unix Support Group.
It appeared in 1977 as part of PWB/UNIX 1.0, the
``Programmer's Work Bench''
derived from
At v6
that was used internally at AT&T.
Both the new binary and old character formats were in use
by 1980, according to the System III source released
by SCO under their
``Ancient Unix''
license.
The character format was adopted as part of
IEEE Std 1003.1-1988 (``POSIX.1'').
XXX when did "newc" appear?  Who invented it?  When did HP come out with their variant?  When did Sun introduce ACLs and extended attributes? XXX
.SH BUGS
.ad l
The
``CRC''
format is mis-named, as it uses a simple checksum and
not a cyclic redundancy check.
.PP
The binary formats are limited to 16 bits for user id, group id,
device, and inode numbers.  They are limited to 16 megabyte and 2
gigabyte file sizes for the older and newer variants, respectively.
.PP
The old ASCII format is limited to 18 bits for
the user id, group id, device, and inode numbers.
It is limited to 8 gigabyte file sizes.
.PP
The new ASCII format is limited to 4 gigabyte file sizes.
.PP
None of the cpio formats store user or group names,
which are essential when moving files between systems with
dissimilar user or group numbering.
.PP
Especially when writing older cpio variants, it may be necessary
to map actual device/inode values to synthesized values that
fit the available fields.
With very large filesystems, this may be necessary even for
the newer formats.


projects/toybox-8.12/toys/posix/cpio.c
/* cpio.c - a basic cpio
 *
 * Copyright 2013 Isaac Dunham <ibid.ag@gmail.com>
 * Copyright 2015 Frontier Silicon Ltd.
 *
 * see https://www.kernel.org/doc/Documentation/early-userspace/buffer-format.txt
 * and http://refspecs.linuxfoundation.org/LSB_4.1.0/LSB-Core-generic/LSB-Core-generic/cpio.html
 * and http://pubs.opengroup.org/onlinepubs/7908799/xcu/cpio.html
 *
 * Yes, that's SUSv2, newer versions removed it, but RPM and initramfs use
 * this archive format. We implement (only) the modern "-H newc" variant which
 * expanded headers to 110 bytes (first field 6 bytes, rest are 8).
 * In order: magic ino mode uid gid nlink mtime filesize devmajor devminor
 * rdevmajor rdevminor namesize check
 * This is the equivalent of mode -H newc in other implementations.
 * We always do --quiet, but accept it as a compatibility NOP.
 *
 * TODO: export/import linux file list text format ala gen_initramfs_list.sh
 * TODO: hardlink support, -A, -0, -a, -L, --sparse
 * TODO: --renumber-archives (probably always?) --ignore-devno --reproducible

USE_CPIO(NEWTOY(cpio, "(ignore-devno)(renumber-inodes)(quiet)(no-preserve-owner)R(owner):md(make-directories)uLH:p|i|t|F:v(verbose)o|[!pio][!pot][!pF]", TOYFLAG_BIN))

config CPIO
  bool "cpio"
  default y
  help
    usage: cpio -{o|t|i|p DEST} [-dLtuv] [--verbose] [-F FILE] [-R [USER][:GROUP] [--no-preserve-owner]

    Copy files into and out of a "newc" format cpio archive.

    -d	Create directories if needed
    -F FILE	Use archive FILE instead of stdin/stdout
    -i	Extract from archive into file system (stdin=archive)
    -L	Follow symlinks
    -o	Create archive (stdin=list of files, stdout=archive)
    -p DEST	Copy-pass mode, copy stdin file list to directory DEST
    -R USER	Replace owner with USER[:GROUP]
    -t	Test files (list only, stdin=archive, stdout=list of files)
    -u	Unlink existing files when extracting
    -v	Verbose
    --no-preserve-owner     Don't set ownership during extract
*/

#define FOR_cpio
#include "toys.h"

GLOBALS(
  char *F, *H, *R;
)

// Read strings, tail padded to 4 byte alignment. Argument "align" is amount
// by which start of string isn't aligned (usually 0, but header is 110 bytes
// which is 2 bytes off because the first field wasn't expanded from 6 to 8).
static char *strpad(int fd, unsigned len, unsigned align)
{
  char *str;

  align = (align + len) & 3;
  if (align) len += (4-align);
  xreadall(fd, str = xmalloc(len+1), len);
  str[len]=0; // redundant, in case archive is bad

  return str;
}

//convert hex to uint; mostly to allow using bits of non-terminated strings
static unsigned x8u(char *hex)
{
  unsigned val, inpos = 8, outpos;
  char pattern[6];

  while (*hex == '0') {
    hex++;
    if (!--inpos) return 0;
  }
  // Because scanf gratuitously treats %*X differently than printf does.
  sprintf(pattern, "%%%dX%%n", inpos);
  sscanf(hex, pattern, &val, &outpos);
  if (inpos != outpos) error_exit("bad hex");

  return val;
}

void cpio_main(void)
{
  int pipe, afd = FLAG(o), reown = !geteuid() && !FLAG(no_preserve_owner),
      empty = 1;
  pid_t pid = 0;
  long Ruid = -1, Rgid = -1;
  char *tofree = 0;

  if (TT.R) {
    char *group = TT.R+strcspn(TT.R, ":.");

    if (*group) {
      Rgid = xgetgid(group+1);
      *group = 0;
    }
    if (group != TT.R) Ruid = xgetuid(TT.R);
  }

  // In passthrough mode, parent stays in original dir and generates archive
  // to pipe, child does chdir to new dir and reads archive from stdin (pipe).
  if (FLAG(p)) {
    if (FLAG(d)) {
      if (!*toys.optargs) error_exit("need directory for -p");
      if (mkdir(*toys.optargs, 0700) == -1 && errno != EEXIST)
        perror_msg("mkdir %s", *toys.optargs);
    }
    if (toys.stacktop) {
      // xpopen() doesn't return from child due to vfork(), instead restarts
      // with !toys.stacktop
      pid = xpopen(0, &pipe, 0);
      afd = pipe;
    } else {
      // child
      toys.optflags |= FLAG_i;
      xchdir(*toys.optargs);
    }
  }

  if (TT.F) {
    int perm = FLAG(o) ? O_CREAT|O_WRONLY|O_TRUNC : O_RDONLY;

    afd = xcreate(TT.F, perm, 0644);
  }

  // read cpio archive

  if (FLAG(i) || FLAG(t)) for (;; empty = 0) {
    char *name, *data;
    unsigned mode, uid, gid, timestamp;
    int test = FLAG(t), err = 0, size = 0, len;

    free(tofree);
    tofree = 0;
    // read header, skipping arbitrary leading NUL bytes (concatenated archives)
    for (;;) {
      if (1>(len = readall(afd, toybuf+size, 110-size))) break;
      if (size || *toybuf) {
        size += len;
        break;
      }
      for (size = 0; size<len; size++) if (toybuf[size]) break;
      memmove(toybuf, toybuf+size, len-size);
      size = len-size;
    }
    if (!size) {
      if (empty) error_exit("empty archive");
      else break;
    }
    if (size != 110 || smemcmp(toybuf, "070701", 6)) error_exit("bad header");
    tofree = name = strpad(afd, x8u(toybuf+94), 110);
    // TODO: this flushes hardlink detection via major/minor/ino match
    if (!strcmp("TRAILER!!!", name)) continue;

    // If you want to extract absolute paths, "cd /" and run cpio.
    while (*name == '/') name++;
    // TODO: remove .. entries

    size = x8u(toybuf+54);
    mode = x8u(toybuf+14);
    uid = (Ruid>=0) ? Ruid : x8u(toybuf+22);
    gid = (Rgid>=0) ? Rgid : x8u(toybuf+30);
    timestamp = x8u(toybuf+46); // unsigned 32 bit, so year 2100 problem

    // (This output is unaffected by --quiet.)
    if (FLAG(t) || FLAG(v)) puts(name);

    if (FLAG(u) && !test) if (unlink(name) && errno == EISDIR) rmdir(name);

    if (!test && FLAG(d) && strrchr(name, '/') && mkpath(name)) {
      perror_msg("mkpath '%s'", name);
      test++;
    }

    // Consume entire record even if it couldn't create file, so we're
    // properly aligned with next file.

    if (S_ISDIR(mode)) {
      if (test) continue;
      err = mkdir(name, mode) && (errno != EEXIST && !FLAG(u));

      // Creading dir/dev doesn't give us a filehandle, we have to refer to it
      // by name to chown/utime, but how do we know it's the same item?
      // Check that we at least have the right type of entity open, and do
      // NOT restore dropped suid bit in this case.
      if (S_ISDIR(mode) && reown) {
        int fd = open(name, O_RDONLY|O_NOFOLLOW);
        struct stat st;

        if (fd != -1 && !fstat(fd, &st) && (st.st_mode&S_IFMT) == (mode&S_IFMT))
          err = fchown(fd, uid, gid);
        else err = 1;

        close(fd);
      }
    } else if (S_ISREG(mode)) {
      int fd = test ? 0 : open(name, O_CREAT|O_WRONLY|O_EXCL|O_NOFOLLOW, mode);

      // If write fails, we still need to read/discard data to continue with
      // archive. Since doing so overwrites errno, report error now
      if (fd < 0) {
        perror_msg("create %s", name);
        test++;
      }

      data = toybuf;
      while (size) {
        if (size < sizeof(toybuf)) data = strpad(afd, size, 0);
        else xreadall(afd, toybuf, sizeof(toybuf));
        if (!test) xwrite(fd, data, data == toybuf ? sizeof(toybuf) : size);
        if (data != toybuf) {
          free(data);
          break;
        }
        size -= sizeof(toybuf);
      }

      if (!test) {
        // set owner, restore dropped suid bit
        if (reown) err = fchown(fd, uid, gid) && fchmod(fd, mode);
        close(fd);
      }
    } else {
      data = S_ISLNK(mode) ? strpad(afd, size, 0) : 0;
      if (!test) {
        err = data ? symlink(data, name)
          : mknod(name, mode, dev_makedev(x8u(toybuf+78), x8u(toybuf+86)));

        // Can't get a filehandle to a symlink or a node on nodev mount,
        // so do special chown that at least doesn't follow symlinks.
        // We also don't chmod after, so dropped suid bit isn't restored
        if (!err && reown) err = lchown(name, uid, gid);
      }
      free(data);
    }

    // Set timestamp.
    if (!test && !err) {
      struct timespec times[2];

      memset(times, 0, sizeof(struct timespec)*2);
      times[0].tv_sec = times[1].tv_sec = timestamp;
      err = utimensat(AT_FDCWD, name, times, AT_SYMLINK_NOFOLLOW);
    }

    if (err) perror_msg_raw(name);

  // Output cpio archive

  } else {
    char *name = 0;
    size_t size = 0;
    unsigned inode = 0;

    for (;;) {
      struct stat st;
      unsigned nlen, error = 0, zero = 0;
      int len, fd = -1;
      char *link = 0;
      ssize_t llen;

      len = getline(&name, &size, stdin);
      if (len<1) break;
      if (name[len-1] == '\n') name[--len] = 0;
      if (!len) continue;
      nlen = len+1;
      if ((FLAG(L)?stat:lstat)(name, &st) || (S_ISREG(st.st_mode)
          && st.st_size && (fd = open(name, O_RDONLY))<0)
          || (S_ISLNK(st.st_mode) && !(link = xreadlink(name))))
      {
        perror_msg_raw(name);
        continue;
      }
      // encrypted filesystems can stat the wrong link size
      if (link) st.st_size = strlen(link);

      if (Ruid>=0) st.st_uid = Ruid;
      if (Rgid>=0) st.st_gid = Rgid;
      if (FLAG(no_preserve_owner)) st.st_uid = st.st_gid = 0;
      if (!S_ISREG(st.st_mode) && !S_ISLNK(st.st_mode)) st.st_size = 0;
      if (st.st_size >> 32) perror_msg("skipping >2G file '%s'", name);
      else {
        if (FLAG(renumber_inodes)) st.st_ino = ++inode;
        if (FLAG(ignore_devno)) st.st_rdev = 0;
        llen = sprintf(toybuf,
          "070701%08X%08X%08X%08X%08X%08X%08X%08X%08X%08X%08X%08X%08X",
          (int)st.st_ino, st.st_mode, st.st_uid, st.st_gid, (int)st.st_nlink,
          (int)st.st_mtime, (int)st.st_size, dev_major(st.st_dev),
          dev_minor(st.st_dev), dev_major(st.st_rdev), dev_minor(st.st_rdev),
          nlen, 0);
        xwrite(afd, toybuf, llen);
        xwrite(afd, name, nlen);

        // NUL Pad header up to 4 multiple bytes.
        llen = (llen + nlen) & 3;
        if (llen) xwrite(afd, &zero, 4-llen);

        // Write out body for symlink or regular file
        if (link) xwrite(afd, link, st.st_size);
        else for (llen = st.st_size; llen; llen -= nlen) {
          nlen = llen > sizeof(toybuf) ? sizeof(toybuf) : llen;
          // If read fails, write anyway (already wrote size in header)
          if (nlen != readall(fd, toybuf, nlen))
            if (!error++) perror_msg("bad read from file '%s'", name);
          xwrite(afd, toybuf, nlen);
        }
        llen = st.st_size & 3;
        if (llen) xwrite(afd, &zero, 4-llen);
      }
      free(link);
      xclose(fd);
    }
    if (CFG_TOYBOX_FREE) free(name);

    // nlink=1, namesize=11, with padding
    dprintf(afd, "070701%040X%056X%08XTRAILER!!!%c%c%c%c", 1, 11, 0, 0, 0, 0,0);
  }
  if (TT.F) xclose(afd);

  if (FLAG(p) && pid) toys.exitval |= xpclose(pid, pipe);
}
